File: 940-fruit-into-baskets/fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int n = fruits.size();
        int l =0, r=0, maxcount =0;
        unordered_map<int,int> mp;
        while(r<n) {
            mp[fruits[r]]++;
            if(mp.size() > 2) {
                mp[fruits[l]]--;
                if(mp[fruits[l]] == 0) mp.erase(fruits[l]);
                l++;
            }
            if(mp.size()<=2) {
                int len = r-l+1;
                maxcount = max(maxcount, len);
            }
            r++;
        }
        return maxcount;
    }
};

File: 940-fruit-into-baskets/fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int n = fruits.length;
        int maxc = Integer.MIN_VALUE;
        HashMap<Integer,Integer> hm = new HashMap<>();
        int l=0, r=0;
        while(r<n) {
            hm.put(fruits[r], hm.getOrDefault(fruits[r],0)+1);
            if(hm.size()>2) {
                if(hm.get(fruits[l]) >1){
                    hm.put(fruits[l],hm.get(fruits[l])-1);
                }
                else hm.remove(fruits[l]);
                l++;
            }
            maxc = Math.max(maxc, r-l+1);
            r++;
        }
        return maxc;

    }
}